Add documentation on how NGINX Agent works and Install instructions (#74)
nginxinc/nginx-amplify-doc: content/faq/nginx-agent.md

---
title: NGINX Agent
description: Questions about NGINX Agent
weight: 20
toc: true
tags: ["docs"]
---

### What Operating Systems are Supported?

> The agent is currently officially packaged and supported for the following Linux flavors only:

Reviewer: Here and throughout, write "NGINX Agent" -- the full product name. We're not allowed to shorten it.

> The agent is currently officially packaged and supported for the following Linux flavors only:

Reviewer: ```suggestion
The NGINX Agent is officially packaged and supported only for the following Linux flavors:
```

>   * Ubuntu 22.04 "jammy" (amd64/arm64)

Reviewer: Could you let me know the source of this list? The NGINX Agent is documented in several places, and it's important that the list of supported distributions is consistent everywhere.

Reviewer: Maybe this is the source: https://docs.nginx.com/nginx-agent/technical-specifications/

If you choose to duplicate content across various documents, be warned: you'll need to be vigilant and resolve any inevitable discrepancies that come up. Do you have a content plan to do this?

Author: @travisamartin The NGINX Agent supports a bigger set of OSes but this list is the OSes on which we have tested the NGINX Agent configured for reporting data to Amplify.

  * Ubuntu 20.04 "focal" (amd64/arm64)
  * Ubuntu 18.04 "bionic" (amd64/arm64)
  * Debian 11 "bullseye" (amd64/arm64)
  * RHEL/CentOS/OEL 8 (amd64/arm64)
  * RHEL/CentOS/OEL 9 (amd64/arm64)
  * Amazon Linux 2 LTS (amd64/arm64)

The following platforms are no longer supported but still can be used with older agent packages powered by Python 2:

  * Ubuntu 16.04 "xenial" (i386/amd64/arm64)
  * Debian 9 "stretch" (i386/amd64)
  * RHEL/CentOS/OEL 6 (i386/amd64)
  * RHEL/CentOS/OEL 7 (amd64)
  * Amazon Linux (amd64)

> Other OS and distributions below are not fully supported yet (and no agent packages are available), however you can grab a specialized install script [here](https://raw.githubusercontent.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh) and see if it works for you. Run [install-nginx-agent.sh](https://raw.githubusercontent.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh) (as root) and follow the dialog. You can copy the API key from the Amplify UI (find it in the **Settings** or in the **New System** pop-up).

Reviewer: ```suggestion
To monitor your infrastructure with NGINX Amplify, install the NGINX Amplify Agent on every system you want to monitor.

The following OS and distributions aren't fully supported yet, and no agent packages are available. However, you can try a specialized install script [here](https://raw.githubusercontent.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh). To use it, run [install-nginx-agent.sh](https://raw.githubusercontent.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh) as root and follow the prompts. You can find the API key in the Amplify user interface, either in the **Settings** or the **New System** pop-up window.
```

> Other OS and distributions below are not fully supported yet (and no agent packages are available), however you can grab a specialized install script [here](https://raw.githubusercontent.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh) and see if it works for you. Run [install-nginx-agent.sh](https://raw.githubusercontent.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh) (as root) instead of *install.sh* and follow the dialog. You can copy the API key from the Amplify UI (find it in the **Settings** or in the **New System** pop-up).

Reviewer 2: Rather than mention this: `instead of *install.sh*`

Should we just exclude it? Will it be more clear and concise if we just say "run this script" such as you already have instead of introducing the idea of another script?

  * FreeBSD
  * SLES
  * Alpine
  * Fedora

> Feel free to [submit](https://github.com/nginx/agent/) an issue or a PR if you find something that has to be fixed.

Reviewer: ```suggestion
If you find something that needs fixing, you're welcome to [submit a PR](https://github.com/nginx/agent/) for the issue.
```

> ### How Do I Start to Monitor My Systems with NGINX Amplify?

Reviewer: This FAQ repeats steps from the install guide. Why not just link? If you duplicate the text, you've got to maintain it in multiple places.

Author: Linked to the Install instructions.

1. Download and run the install script.

  ```bash
    curl -sS -L -O \
    https://github.com/nginxinc/naas-agent/blob/main/packages/install-nginx-agent.sh && \
    API_KEY='YOUR_API_KEY' sh ./install-nginx-agent.sh
  ```

>    where YOUR_API_KEY is a unique API key assigned when you create an account with NGINX Amplify. You can also find the API key in the **Account** menu.

Reviewer 2: You have some extra space before: `where YOUR_API_KEY is`

> 2. Verify that the Agent has started.

Reviewer: ```suggestion
2. Verify NGINX Agent has started.
```

   ```bash
   ps ax | grep -i 'nginx-agent'
   3725474 ?        Ssl    0:00 /usr/bin/nginx-agent
   ```

> For manual installation, please check the [user guide]({{< relref "/install-manage-nginx-agent/installing-agent#installing-the-agent-manually" >}}).

Reviewer: ```suggestion
For manual installation, refer to the [user guide]({{< relref "/install-manage-nginx-agent/installing-agent#installing-the-agent-manually" >}}).
```

### What Do I Need to Configure the NGINX Agent to Report Metrics Correctly?

> After you install and start the agent, it should start reporting right away, pushing aggregated data to the NGINX Amplify services at regular 1-minute intervals. It'll take about a minute for the new system to appear in the Amplify web interface.

Reviewer: ```suggestion
Once you install and launch NGINX Agent, it will immediately begin reporting, sending aggregated data to NGINX Amplify services every minute. Expect the new system to show up in the Amplify web interface within about a minute.
```

> After you install and start the agent, it should start reporting right away, pushing aggregated data to the Amplify backend at regular 1-minute intervals. It'll take about a minute for the new system to appear in the Amplify web interface.

Reviewer 2: `to the Amplify backend` -> `to the NGINX Amplify services`

> If you don't see the new system or NGINX in the web interface, or (some) metrics aren't being collected, please check the following:

Reviewer: ```suggestion
If the new system or NGINX doesn't appear in the web interface, or if some metrics are missing, refer to the following steps:
```

> 1. The Nginx Agent package has been successfully [installed]({{< relref "/install-manage-nginx-agent" >}}), and no warnings were shown during the installation.

Reviewer 2: `Nginx` -> `NGINX`

8. All NGINX configuration files are readable by the agent user ID (check owner, group, and permissions).

9. The system DNS resolver is correctly configured, and *receiver-grpc.amplify.nginx.com* can be successfully resolved.

10. *selinux(8)*, *apparmor(7)* or [grsecurity](https://grsecurity.net) are not interfering with the metric collection. E.g. for *selinux(8)* check **/etc/selinux/config**, try `setenforce 0` temporarily and see if it improves the situation for certain metrics.

11. Some VPS providers use hardened Linux kernels that may restrict non-root users from accessing */proc* and */sys*. Metrics describing system and NGINX disk I/O are usually affected. There is no easy workaround except for allowing the agent to run as `root`. Sometimes fixing permissions for */proc* and */sys/block* may work.

### How Do I Verify that NGINX Agent Is Correctly Installed?

1. On Ubuntu/Debian use:

   ```bash
   dpkg -s nginx-agent
   ```

2. On CentOS and Red Hat use:

   ```bash
   yum info nginx-agent
   ```

> ### How Can I Update NGINX Agent?

Reviewer: Why repeat the steps in the Updating NGINX Agent guide? Consider linking instead. Otherwise, you'll need to maintain these instructions twice.

Author: Linked to the update instructions.

1. On Ubuntu/Debian use:

   ```bash
   apt-get update && \
   apt-get install nginx-agent
   ```

2. On CentOS use:

   ```bash
   yum makecache && \
   yum update nginx-agent
   ```

### What System Resources are Required?

> Under 10% of the CPU and a few dozen MBs of RSS memory will be consumed. If you notice any anomalies in the system resource consumption, please fill in a support ticket through Intercom.

Reviewer: ```suggestion
Expect to use under 10% of the CPU and a few dozen MBs of RSS memory. If you notice any unusual resource consumption, submit a support ticket through Intercom.
```

### How Do I Restart NGINX Agent?

   ```
   # service nginx-agent restart
   ```

> ### How Can I Uninstall NGINX Agent?

Reviewer: Why does this FAQ simply repeat the steps from the Uninstall NGINX Agent guide? Now, you have to maintain these steps in two places. Perhaps linking to the guide would be more efficient?

> We work hard to make the agent consume as little memory and CPU as possible. Usually, it should be under 10% of CPU usage and a few dozen MBs of RSS memory. If you notice any anomalies in the system resource consumption, please fill in a support request through https://my.f5.com, please attach the debug log to the support case. 

Reviewer 2: Don't need the following: `We work hard to make the agent consume as little memory and CPU as possible.`

`Usually, it should be under 10% of CPU usage and a few dozen MBs of RSS memory.` -> `Under 10% of the CPU and a few dozen MBs of RSS memory will be consumed.`

1. On Ubuntu/Debian use:

   ```bash
   apt-get remove nginx-agent
   ```

2. On CentOS and Red Hat use:

   ```bash
   yum remove nginx-agent
   ```

### How Can I Override System Hostname?

> If the agent is not able to determine the system's hostname, you can define it manually in **/etc/nginx-agent/nginx-agent.conf**

Reviewer: ```suggestion
If NGINX Agent can't determine the system's hostname, manually define it in `/etc/nginx-agent/nginx-agent.conf`.
```

Author: Removed this section as it is not correct for NGINX Agent

Find the following section, and fill in the desired hostname:

```nginx
[credentials]
..
hostname = myhostname1
```

> The hostname should be valid — the following aren't valid hostnames:

Reviewer: ```suggestion
The hostname must be valid. The following are not valid hostnames:
```

Author: Removed this section as it is not correct for NGINX Agent

  * localhost
  * localhost.localdomain
  * localhost6.localdomain6
  * ip6-localhost

### Can I Use NGINX Agent with Docker?

> Please check the [following section](https://docs.nginx.com/nginx-agent/docker-images/) of the agent repository to find out more. Keep in mind that the support for a Docker environment is currently experimental.

Reviewer: ```suggestion
For details, refer to the [Docker Images](https://docs.nginx.com/nginx-agent/docker-images/) topic in the NGINX Agent documentation. Note that Docker environment support is currently experimental.
```
